apps/crn-server/src/data-providers/contentful/working-groups.data-provider.ts

```diff
@@ -0,0 +1,253 @@
+} from '@asap-hub/model';
+import {
+  patchAndPublish,
+  GraphQLClient,
+  FETCH_WORKING_GROUPS,
+  FETCH_WORKING_GROUP_BY_ID,
+  FetchWorkingGroupByIdQuery,
+  FetchWorkingGroupByIdQueryVariables,
+  FetchWorkingGroupsQuery,
+  FetchWorkingGroupsQueryVariables,
+  Environment,
+  parseRichText,
+  RichTextFromQuery,
+  WorkingGroupsFilter,
+  addLocaleToFields,
+  Link,
+} from '@asap-hub/contentful';
+
+import { WorkingGroupDataProvider } from '../types';
+import { parseContentfulGraphqlCalendarToResponse } from '../../entities';
+
+export type WorkingGroupItem = NonNullable<
+  NonNullable<
+    FetchWorkingGroupsQuery['workingGroupsCollection']
+  >['items'][number]
+>;
+
+export class WorkingGroupContentfulDataProvider
+  implements WorkingGroupDataProvider
+{
+  constructor(
+    private contentfulClient: GraphQLClient,
+    private getRestClient: () => Promise<Environment>,
+  ) {}
+
+  async fetch(
+    options: FetchWorkingGroupOptions,
+  ): Promise<WorkingGroupListDataObject> {
+    const { search, take = 10, skip = 0, filter } = options;
+
+    const where: WorkingGroupsFilter = {};
+    const words = (search || '').split(' ').filter(Boolean); // removes whitespaces
+
+    if (words.length) {
+      const filters: WorkingGroupsFilter[] = words.reduce(
+        (acc: WorkingGroupsFilter[], word: string) =>
+          acc.concat([
+            {
+              OR: [
+                { title_contains: word },
+                { description_contains: word },
+                { shortText_contains: word },
+              ],
+            },
+          ]),
+        [],
+      );
+      where.AND = filters;
+    }
+
+    if (filter && filter.complete !== undefined) {
+      where.AND = [...(where.AND || []), { complete: filter.complete }];
+    }
+
+    const { workingGroupsCollection } = await this.contentfulClient.request<
+      FetchWorkingGroupsQuery,
+      FetchWorkingGroupsQueryVariables
+    >(FETCH_WORKING_GROUPS, {
+      limit: take,
+      skip,
+      where,
+    });
+
+    if (!workingGroupsCollection?.items) {
+      return {
+        total: 0,
+        items: [],
+      };
+    }
+
+    return {
+      total: workingGroupsCollection?.total,
+      items: workingGroupsCollection?.items
+        .filter((x): x is WorkingGroupItem => x !== null)
+        .map(parseContentfulGraphQlWorkingGroup),
+    };
+  }
+
+  async fetchById(id: string): Promise<WorkingGroupDataObject | null> {
+    const { workingGroups } = await this.contentfulClient.request<
+      FetchWorkingGroupByIdQuery,
+      FetchWorkingGroupByIdQueryVariables
+    >(FETCH_WORKING_GROUP_BY_ID, { id });
+
+    if (!workingGroups) {
+      return null;
+    }
+
+    return parseContentfulGraphQlWorkingGroup(workingGroups);
+  }
+
+  async update(
+    id: string,
+    update: WorkingGroupUpdateDataObject,
+  ): Promise<void> {
+    const environment = await this.getRestClient();
+    const workingGroup = await environment.getEntry(id);
+
+    await Promise.all(
```

Comment:
Do we need to add rate limiting here?

Reply:
I believe we don't send a lot of update requests in sequence, but I've added it now to be safe: [review-suggestion: add limiter](https://github.com/yldio/asap-hub/pull/3110/commits/b6da28df0028b2d79c4f248cc2e0aa89f514567b)